### DISCUSSION TITLE
How to use on Nginx Reverse Proxy?

### DISCUSSION THREAD

**Discussion by satonotdead** [ItzCrazyKns/Vane#477] [General]
Coming from #62 #69 #142 #154 #241 #244 #318 #467 and after trying by months, I just can't get Perplexica working behind Nginx reverse proxy.

Can someone share their current working templates of `docker-compose.yml` and `nginx.conf`?

**Comment by bpawnzZ** [comment]
Inside nginx.conf

```
server {
    listen 80;
    server_name searxng.domain.com;
    return 301 https://$host$request_uri; # Redirect all HTTP to HTTPS
}

server {
    listen 443 ssl;
    server_name searxng.domain.com;

    ssl_certificate /etc/letsencrypt/live/searxng.domain.com/fullchain.pem;
    ssl_certificate_key /etc/letsencrypt/live/searxng.domain.com/privkey.pem;

    location / {
        proxy_pass http://localhost:4000; # Assuming the searxng service is available on port 4000
        proxy_set_header Host $host;
        proxy_set_header X-Real-IP $remote_addr;
        proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
        proxy_set_header X-Forwarded-Proto https;
    }
}

server {
    listen 80;
    server_name perplexica.domain.com;
    return 301 https://$host$request_uri; # Redirect all HTTP to HTTPS
}

server {
    listen 443 ssl;
    server_name perplexica.domain.com;

    ssl_certificate /etc/letsencrypt/live/perplexica.domain.com/fullchain.pem;
    ssl_certificate_key /etc/letsencrypt/live/perplexica.domain.com/privkey.pem;

    # Proxy pass for the API
    location /api {
        proxy_pass http://localhost:3001; # Assuming the backend service is available on port 3001
        proxy_set_header Host $host;
        proxy_set_header X-Real-IP $remote_addr;
        proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
        proxy_set_header X-Forwarded-Proto https;
    }

    # Proxy pass for the frontend
    location / {
        proxy_pass http://localhost:3000; # Assuming the frontend service is available on port 3000
        proxy_set_header Host $host;
        proxy_set_header X-Real-IP $remote_addr;
        proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
        proxy_set_header X-Forwarded-Proto https;
    }
}
```
where i did localhost in above example for pointing to services your could also use internal IPs as long as everything is on the same network. The API is really the only tricky part.  I'm currently using traefik.. I find it simpler and just like it more but the above should be good to go for NGINX

Also.. To get this working in Traefik atleast but likely nginx also.  in my docker-compose modify yours to be equivilent to below

```
- NEXT_PUBLIC_API_URL=https://perplexica.domain.com/api
- NEXT_PUBLIC_WS_URL=wss://perplexica.domain.com/api
 ```
 the /api is for some reason necessary to be on both.
 Then do a clean build and your good to go.

clean build w/
```
docker compose build --no-cache
docker compose up -d
```

**Comment by satonotdead** [comment]
Thanks @ms-jpq, I tried your config and now I see a different error:

`can’t establish a connection to the server at wss://perp.mydomain.tld/?chatModel=gpt-3.5-turbo&chatModelProvider=openai&embeddingModel=text-embedding-3-small&embeddingModelProvider=openai`

Tried with and without `/api` on `NEXT_PUBLIC_WS_URL`. What can be wrong on my instance? I'm using the same like you and ``--build`` on docker plus deleted data/db files.

**Reply by ms-jpq** [reply]
one reason could be that if you are not on https you should be using ws instead of wss.

**Reply by kjbrum** [reply]
I'm following this approach and running into the exact same issue. I haven't set up https yet and made sure I'm using `ws` instead of `wss`, still getting the websocket error. 🤔

**Reply by satonotdead** [reply]
> one reason could be that if you are not on https you should be using ws instead of wss.

I'm using SSL on the domains proxied by Nginx and I tried `http://` and `ws://` but no luck.

**Comment by kjbrum** [comment]
@satonotdead I got mine working with the following:

This comment helped: https://github.com/ItzCrazyKns/Perplexica/issues/142#issuecomment-2140916344

## `nginx.conf`
```sh
map $http_upgrade $connection_upgrade {
    default upgrade;
    ''      close;
}

upstream perplexica-frontend {
    server localhost:3000;
}

upstream perplexica-backend {
    server localhost:3001;
}

# Redirect all HTTP to HTTPS
server {
    listen 80;
    listen [::]:80;

    server_name perplexica.domain.com;

    return 301 https://$host$request_uri;
}

############################
# perplexica.domain.com #
############################
server {
    listen 443 ssl http2;
    listen [::]:443 ssl http2;

    server_name perplexica.domain.com;

    allow <ip_address(es)_to_allow_access>;
    deny all;

    # SSL configuration    
    ssl_certificate /etc/letsencrypt/live/perplexica.domain.com/fullchain.pem;
    ssl_certificate_key /etc/letsencrypt/live/perplexica.domain.com/privkey.pem;
    include /etc/letsencrypt/options-ssl-nginx.conf;
    ssl_dhparam /etc/letsencrypt/ssl-dhparams.pem;

    # Proxy pass for the backend/API
    location /api {
        proxy_pass http://perplexica-backend;
        proxy_set_header Host $host;
        proxy_set_header X-Real-IP $remote_addr;
        proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
        proxy_set_header X-Forwarded-Proto https;
    }

    # Proxy pass for websockets
    location /ws {
        proxy_pass http://perplexica-backend;
        #include proxy_params;
        proxy_set_header Host $host;
        proxy_set_header X-Real-IP $remote_addr;
        proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
        proxy_set_header X-Forwarded-Proto $scheme;
        
        # Important!
        proxy_http_version 1.1;
        proxy_set_header Upgrade $http_upgrade;
        proxy_set_header Connection $connection_upgrade;
    }

    # Proxy pass for the frontend
    location / {
        proxy_pass http://perplexica-frontend;
        proxy_set_header Host $host;
        proxy_set_header X-Real-IP $remote_addr;
        proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
        proxy_set_header X-Forwarded-Proto https;
    }
}
```

## `docker-compose.yaml`
```sh
- NEXT_PUBLIC_API_URL=https://perplexica.domain.com/api
- NEXT_PUBLIC_WS_URL=wss://perplexica.domain.com/ws
```

**Comment by satonotdead** [comment]
Thanks @kjbrum, I tried this configuration and I see this error on console:

`The Same Origin Policy disallows reading the remote resource at http://127.0.0.1:3001/api/models. (Reason: CORS request did not succeed)`

Curl show status 200 onlocalhost but I can't still get it through Nginx.

**Reply by realies** [reply]
I managed to get around this bug by running `yarn build` inside the frontend docker container and restarting it.

**Reply by realies** [reply]
@ItzCrazyKns, the ARG default values get baked into the frontend bundle.
Setting the environment variables does not change them.

**Comment by bpawnzZ** [comment]
Just somehow noticed why this didn't work for yall.. Use this fork.  It works with above...

```
https://github.com/rqi14/Perplexica-docker
```

Just had to get this project back up and going and just went through it all.  I was using Traefik but pretty sure that will solve your problems also

Also in your compose put /api after both ws and api 

```
args:
        - NEXT_PUBLIC_API_URL=https://perplexica.domain.com/api
        - NEXT_PUBLIC_WS_URL=wss://perplexica.domain.com/api
        ```

**Reply by realies** [reply]
Maybe the author could bother to fix the usage with their Docker image too.

**Comment by satonotdead** [comment]
I wonder if that should be updated on the main app.

**EDIT:** it seems the last version of Perplexica fix the general issue around here but the current Nginx template could be also updated.
